### PR TITLE
feat: add erc721

### PR DESCRIPTION
**Pull request by ftupas**
This PR implements the following:

1. An ERC721 token that can only be minted by the cohort contract address
2. An **external** function `mint_basecamp_token` students to claim their graduation token once they've completed all the exercises
3. Tests for the `Starknet Basecamp Token` contract and minting function

Please let me know what you think and if there's anything else I can add, thanks! 🙂

### CODE REVIEW

File: src/tokens/basecamp/library.cairo [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+        # thus, each token is guaranteed to be a unique uint256
+        let (new_balance: Uint256, _) = uint256_add(balance, Uint256(1, 0))
+        ERC721_balances.write(to, new_balance)
+
+        # low + high felts = uint256
+        ERC721_owners.write(token_id, to)
+
+        # Emit Transfer event
+        Transfer.emit(_from=0, to=to, tokenId=token_id)
+        return ()
```
Comment (ftupas) [inline src/tokens/basecamp/library.cairo:388]: Entrypoint here. This is the core logic of the token minting i.e. only the cohort contract address can call this external function. Also added a pausing functionality wherein the moderator could potentially do a bunch of admin work like changing the URI or the cohort contract address when the this contract is paused

File: src/tokens/basecamp/metadata.cairo [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+# Storage
+#
+
+@storage_var
+func ERC721_token_uri(index: felt) -> (res: felt):
+end
+
+@storage_var
+func ERC721_token_uri_len() -> (res: felt):
+end
```
Comment (ftupas) [inline src/tokens/basecamp/metadata.cairo:26]: This is to store the metadata tokenURI when it's longer than 31 characters

File: src/tokens/basecamp/metadata.cairo [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.cairo_builtins import HashBuiltin, SignatureBuiltin
+from starkware.cairo.common.uint256 import Uint256
+
+from openzeppelin.token.erc721.library import _exists
+
+from openzeppelin.introspection.ERC165 import ERC165
+
+from src.utils.ShortString import uint256_to_ss
+from src.utils.Array import concat_arr
```
Comment (ftupas) [inline src/tokens/basecamp/metadata.cairo:14]: These are helper functions that I borrowed and modified from https://github.com/l-henri/starknet-erc721

File: src/basecamp_cohort_0.cairo [hunk 4]
```diff
@@ -105,6 +115,66 @@ func student_register{syscall_ptr:felt*, pedersen_ptr: HashBuiltin*, range_check
...
+@external
+func set_starknet_basecamp_address{syscall_ptr:felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(address: felt):
+    only_moderator()
+    starknet_basecamp_address_.write(address)
+    return ()
+end
+
+@external
+func mint_basecamp_token{syscall_ptr:felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+):
```
Comment (ftupas) [inline src/basecamp_cohort_0.cairo:127]: Student can mint a graduation token if they've completed the course, in addition to other checks as seen below. This calls the ERC721 contract's `mint` function and mints the token to the student (caller)

**Comment by chee-chyuan** [comment]
Hey I love it!!. one small question, i see the proxy library is being used and an implementation is also being set however i dont see where are we using the implementation. maybe i missed it but yea not being too nit picky about this. merging this as it looks awesome!!!!

**Comment by ftupas** [comment]
Thanks for the review @chee-chyuan ! I was thinking to add an upgradeability functionality but had to rethink how to implement it and forgot to put them in another commit 😅 Perhaps we can have discussions on how to approach this 🙏

**Pull request merged**